### DISCUSSION TITLE
Where can I find examples of this library?

### DISCUSSION THREAD

**Discussion by mndrake** [caioproiete/exceldna-addin-abstractions#1] [Q&A]
I found this library via NuGet, but have been unable to use this library to mock Excel-DNA. I was hoping that I could use this to unit test methods that used the built in Excel-DNA async methods. Any examples or sample test projects that you could direct me to would be appreciated.

Thanks,

Dave

**Comment by darul75** [comment]
same here ;)

**Comment by mistresseve666** [comment]
As posted in Excel-DNA/ExcelDna#21, would be great to see an example AddIn that uses these wrappers and possibly even highlights using IoC (as best as it can be using an Implementation AddIn) with ExcelDNA.

**Comment by augustoproiete** [comment]
Thanks guys. It's been on my queue for ages now, but haven't had the time yet. Maybe one (or some) of you could give me a hand to accelerate this? i.e. Send PRs with examples of regular add-ins with (broken or commented out) unit tests that show what and how you'd like to test on these sample add-ins, and I can convert them to use [ExcelDna.Abstractions](https://github.com/augustoproiete/exceldna-abstractions/)? /cc @mndrake, @darul75, @robodude666

## 

@robodude666 I like the part where you say "as best as it can be" because whilst it is possible to get some level of DI with something like [ExcelDna.Abstractions](https://github.com/augustoproiete/exceldna-abstractions/), it is not as pretty as one would like.

Unfortunately ExcelDna makes it impossible to have proper constructor injection, due to the lifetime of the objects (e.g. `IExcelAddIn.AutoOpen` is not a great entry point, as all ribbons get instantiated first, even before `IExcelAddIn.AutoOpen` runs, all formulas as static, etc.).

**Comment by zoll22** [comment]
Did anyone one happen to make any examples, would be really helpful to see some uses.

**Comment by zedr0n** [comment]
Can you please elaborate why you think it's impossible to support constructor injection? 

Before I've seen your project I've already written my own [wrapper](https://github.com/zedr0n/Addin.Packages) and I do all the injection in AutoOpen which seems to work fine. I've also found the requirement to have functions static too onerous so I actually used expression trees to convert the member functions to static ( see [here](https://github.com/zedr0n/Addin.Packages/blob/master/Excel.Addin.Common/PublicRegistration.cs) ) on the fly and using a dictionary service which links internal objects to a 'handle' string which represent the object in Excel. 

Roughly, a static UDF function is exported  f(obj_handle, params) where the underlying lambda expression tree will call dictionary service to convert obj_handle string to object and then invoke the required member method on the converted object.

**Comment by augustoproiete** [comment]
Hey @zedr0n, as I said above, the issue is to use construction injection with Ribbons. All ribbons are constructed before `AutoOpen` is called, so in order to have construction injection on Ribbons, we can't use ExcelDna's implementation as is... We'd have to roll out own own implementation, and instantiate it ourselves in `AutoOpen`.
